gnumed/gnumed/client/doc/man-pages/gm-fingerprint_db.8
.TH GM-FINGERPRINT_DB 8 "2011 Oct 22nd" "GNUmed server fingerprint database"

.SH NAME
.B gm-fingerprint_db
- GNUmed server fingerprint database

.SH SYNOPSIS
.B gm-fingerprint_db
.I <target database>
.I <gm-dbo password>
.I [--with-dump]

.I <target database>:
a GNUmed database (such as "gnumed_VERSION")

.I <gm-dbo password>
the password for the database owner <gm-dbo>

.I [--with-dump]
if this option is given a dump of the database schema structure
will be included created by:

	SELECT gm.concat_table_structure();

.SH DESCRIPTION
The
.B gm-fingerprint_db
script can be used to fingerprint a GNUmed database
in order to aid identifying particular databases
or verifying database transfers (not upgrades).

The fingerprint of the database is written to the file

.B 	gm_db-<target database>-fingerprint.log

in the current directory.

.SH EXAMPLES

Fingerprint the database gnumed_v20:

.B 	gm-fingerprint_db
gnumed_v20 secret-password

Include a database schema structure dump:

.B 	gm-fingerprint_db
gnumed_v20 secret-password --with-dump

.SH AUTHOR
This manual page was written by Karsten Hilbert <karsten.hilbert@gmx.net>.


CHANGELOG
*****************************************************************
# client
*****************************************************************

------------------------------------------------
# rel-0-7-patches
------------------------------------------------

	0.7.0

- new: manage diagnostic orgs
- new: manage date of death per patient
- new: manage discontinuation of substance intake
- new: turn substance intake entry into allergy
- new: hooks
	- after_substance_intake_modified
	- hook after_test_result_modified
- new: visual progress notes
- new: initial support for FreeDiams drug database frontend

- improved: login dialag:
	- hint about local DB on failure
	- do not show password ;-)
- improved: show allergies in substance intake edit area
- improved: warn on creating database users with empty passwords
- improved: default region/country for new-person creation
- improved: show lab name/contact details in measurement tooltip
- improved: always use new person creation rather than old
- improved: placeholders:
	- name::*::*
- improved: various typos
- improved: do not fail if unable to show part when adding documents
- improved: cleaner display of documents in tree
- improved: ARRIBA is now v2.0.4
- improved: using Git for SCM now
- improved: gnumed.py now directly under client/

------------------------------------------------
# rel-0-6-patches
------------------------------------------------

	0.6.3

FIX: faulty tooltip on EMR tree items w/ no data to display

	0.6.2

FIX: properly attribute James Busser, MD
FIX: do not fail on attempt to <delete> when no allergy entry selected
FIX: do not fail on tooltipping empty results/meds grids
FIX: fail less haphazardly on invalid dates in wx -> py conversion

	0.6.1

FIX: WRONG most-recent result displayed in measurement type tooltip
FIX: missing .GetValue() on abnormality indicator PRW on updating a test result
FIX: exception on problem with any placeholder rather than continuing with a notice
FIX: work around one more bogus PyDeadObjectError exception
FIX: re-add pt_BR translation to release tarball
FIX: missing dependencies/faulty paths in gm-install_client_locally.sh
FIX: exception on adding health issue from GUI
FIX: several typos

	0.6.0

- can promote an episode to being a health issue
- can add home phone/external ID to newly created patient directly
- can track diagnostic certainty classification (ABCD) on episodes and issues
- can track procedures performed on a patient
- can do end user friendly free-text search across all EMRs
- can move all progress notes of a pre-selected list of encounters to another episode
- can manage provinces
- can manage substance intake
- can print medication list
- can print LaTeX as well as OOo letters
	- referral letter template contributed by Rogerio Luz and James Busser
- can interface with German "MMI/Gelbe Liste" external drug database
	- show info on drug/substance by PZN / name
	- show interactions
	- import drugs/substances
- can display UI in Polish and others (partially)
- can include potential problems in problem list of soap plugin
- can remove DOB from person

- improved (more) placeholders
	- gender to re placement mapper
	- medication list
	- allergies list
	- problem list
- improved inbox

- improved tarballs: include schema/API docs, better names
- improved import path detection

------------------------------------------------
# rel-0-5-patches
------------------------------------------------

	0.5.1 -> 0.5.2

FIX: obsolete LogException in gmMimeLib.py
FIX: failure on adding patient to waiting list with no patient active
FIX: wx.K_* must be wx.WXK_*
NEW: better default size of encounter details editor

	0.5.0 -> 0.5.1

FIX: be more careful when deleting measurement types - they could be in use
FIX: faulty access to "pk" column after creating measurement type from phrasewheel
FIX: missing "import os" in measurement widgets code
FIX: improved starting of new encounter from SOAP widget

NEW: install ARRIBA jar into /usr/local/bin
NEW: man page for gm-install_arriba

	0.5.0

- can track hospital stays
- can group health issues by arbitrary grouping tag
- can display/edit *all* encounters of a patient in a list
- can move individual soap rows between encounters
- can delete progress notes
- can delete workplaces from GUI
- can show measurements encyclopedia from grid
- can manage measurement types
- can list meta measurement types
- can start external MI/stroke risk assessment calculator
- can install ARRIBA MI/stroke risk assessment calculator
- can download and import LOINC data
- can import ATC data
- can disconnect from database welcome message dialog
- can handle persons without DOB
- can display deceased person's age properly
- can remove persons from the database

- new/simpler placeholder style for OOo documents
- protection against %s-count mismatch in _()
- improved display of measurement types in grid
- improved sending of bug reports (log = attachment)
- improved new-patient dialog
- improved sorting of documents in formatted episode summary
- improved "About Database"/startup banner
- improved episode formatting

- has some revision info in EMR Journal view
- has built-in fallback workplace for problem situations

------------------------------------------------
# rel-0-4-patches
------------------------------------------------

	0.4.6 -> 0.4.7

FIX: faulty get_selected_cells() in measurement grid
FIX: failed to properly check problem field in measurements edit area
FIX: missing ) on setting document type translation

	0.4.5 -> 0.4.6

FIX: missing space in update-available message
FIX: search_for_document by type ID: fk_type must be pk_type

	0.4.4 -> 0.4.5

FIX: _log.panic hiccuping on .conf files with malformed lists, should be _log.critical
FIX: on Python 2.6 __new__ doesn't like extra parameters anymore

	0.4.3 -> 0.4.4

FIX: re-enable path sanity check that fell off when fixing Windows
FIX: properly backpatch "properly find last-but-one encounter for an episode or issue"

gnumed/gnumed/server/gm-fingerprint_db.py
#!/usr/bin/env python

#==============================================================
# This script can be used to fingerprint a GNUmed database.
#
# usage: ./gm-fingerprint_db.py <database name> <gm-dbo password>
#
# author: Karsten Hilbert
# license: GPL v2 or later
#
#==============================================================

import sys
import psycopg2
import io

database = sys.argv[1]
passwd = sys.argv[2]
dsn = u'dbname=%s user=gm-dbo password=%s' % (database, passwd)
queries = [
	("SELECT pg_size_pretty(pg_database_size('%s'))" % database, "Size (DB)"),
	("SELECT md5(gm.concat_table_structure())", "Schema hash"),
	("SELECT setting FROM pg_settings WHERE name = 'server_version'", "Version (PG)"),
	("SELECT setting FROM pg_settings WHERE name = 'server_encoding'", "Encoding (PG)"),
	("SELECT setting FROM pg_settings WHERE name = 'lc_collate'", "LC_COLLATE (PG)"),
	("SELECT setting FROM pg_settings WHERE name = 'lc_ctype'", "LC_CTYPE (PG)"),
	("SELECT count(1) FROM dem.identity", "Patients"),
	("SELECT count(1) FROM clin.encounter", "Contacts"),
	("SELECT count(1) FROM clin.episode", "Episodes"),
	("SELECT count(1) FROM clin.health_issue", "Issues"),
	("SELECT count(1) FROM clin.test_result", "Results"),
	("SELECT count(1) FROM clin.vaccination", "Vaccinations"),
	("SELECT count(1) FROM blobs.doc_med", "Documents"),
	("SELECT count(1) FROM blobs.doc_obj", "Objects"),
	("SELECT count(1) FROM dem.org", "Organizations"),
	("SELECT count(1) FROM dem.org_unit", "Organizational units"),
	("SELECT max(modified_when) FROM audit.audit_fields", "Last .modified_when"),
	("SELECT max(audit_when) FROM audit.audit_trail", "Last .audit_when")
]

fname = u'gm_db-%s-fingerprint.log' % database
#==============================================================
outfile = io.open(fname, mode = 'wt', encoding = 'utf8')

outfile.write(u"Fingerprinting GNUmed database ...\n")
outfile.write(u"\n")
outfile.write(u"%20s: %s\n" % (u"Name (DB)", database))

conn = psycopg2.connect(dsn=dsn)
curs = conn.cursor()

for cmd, label in queries:
	curs.execute(cmd)
	rows = curs.fetchall()
	outfile.write(u"%20s: %s\n" % (label, rows[0][0]))

if len(sys.argv) > 3:
	if sys.argv[3] == '--with-dump':
		curs.execute('SELECT gm.concat_table_structure()')
		rows = curs.fetchall()
		outfile.write(u"\n%s\n" % rows[0][0])

curs.close()
conn.rollback()

outfile.close()
#==============================================================
